### ISSUE TITLE
Port twisted/internet/abstract.py to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5980]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5980|
|Type|enhancement|
|Created|2012-09-14 13:27:42Z|
|Branch|https://github.com/twisted/twisted/tree/abstract-python3-5980-3|

Basis of all reactor implementations.  Needed to get very far on those.

<details><summary>Searchable metadata</summary>

```
trac-id__5980 5980
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_abstract_python3_5980_3 branches/abstract-python3-5980-3
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347629262000000 1347629262000000
changetime__1347892193000000 1347892193000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [35649]) Branching to 'abstract-python3-5980'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [35669]) Branching to 'abstract-python3-5980-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

Maybe not 100% working yet, but better covered now, the marked tests do pass, and at least the thorny issue of `buffer` is kinda addressed.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/abstract-python3-5980-2)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set owner to @exarkun|
|-|-|

  1. I assume you removed the doWrite return value of 0 on purpose? That does mean the docstring is no longer accurate; it should be updated.
  1. Did you make sure no reactor relies on that return value? (I hope not...)
  1. `memoryview` (also available in 2.7), seems like a suitable replacement for `buffer`. If it is, file a ticket to switch to using that when available, and then we don't need to have Python 3-specific code. Or just do it here, whichever is easier. Possibly separate ticket is better, since we'd want to check if it's significantly slower on 2.7.
  1. Minor merge conflict; "solvable" by point below, maybe.
  1. I feel that we shouldn't be adding `twisted.internet.abstract` to ported list until we've actually ported a few more test modules and have a bit more assurance it works. Only really important if other people end up working on this, but seems like good idea to be clear about what is _really_ ported.

(As a side note, we should start exploring using bytearrays now that we've upgraded minimal requirement to 2.6.)

Please address the above, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

> I assume you removed the doWrite return value of 0 on purpose? That does mean the docstring is no longer accurate; it should be updated.

Yes.  Okay.  r35726.

> Did you make sure no reactor relies on that return value? (I hope not...)

I can't check all reactors, of course.  None of our reactors depend on the difference between 0 and None though, and the interface specifies None be returned.

> memoryview (also available in 2.7), seems like a suitable replacement for buffer. If it is, file a ticket to switch to using that when available, and then we don't need to have Python 3-specific code. Or just do it here, whichever is easier. Possibly separate ticket is better, since we'd want to check if it's significantly slower on 2.7.

memoryview is apparently useless for this case - http://bugs.python.org/issue15945

> I feel that we shouldn't be adding twisted.internet.abstract to ported list until we've actually ported a few more test modules and have a bit more assurance it works. Only really important if other people end up working on this, but seems like good idea to be clear about what is _really_ ported.

Sure.  r35727.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [35729]) Branching to 'abstract-python3-5980-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [35733]) Merge abstract-python3-5980-3

Author: exarkun
Reviewer: itamar
Fixes: [#5980](https://github.com/twisted/twisted/issues/5980)

Some basic steps towards having `twisted.internet.abstract` work on Python 3.  Syntax changes
and missing test coverage is the focus here.  Some simplification of the `doWrite` implementation,
too.
